### ISSUE TITLE
PIDFID breaks pkexec with ConsoleKit session tracking

### ISSUE THREAD

**Issue by 0strodamus** [polkit-org/polkit#451]
Compiling polkit 124 with ConsoleKit 1.26 and polkit option session_tracking=ConsoleKit breaks launching any application with pkexec.
The following error is generated on the command line:
> Error checking for authorization org.freedesktop.policykit.exec: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Process not found.

Reverting the PIDFID commits starting with [polkitd: use PIDFDs if available to track processes](https://github.com/polkit-org/polkit/commit/7f0d792323cdca70b3d581cc9ed54df3d844a637) and related subsequent commits fixes the issue.

Also commenting or removing the HAVE_PIDFD_OPEN line from meson.build will fix the issue.

Does this functionality require elogind or systemd? Can an option be added to meson_options.txt to toggle HAVE_PIDFD_OPEN off for systems that it isn't going to work on?

**Comment by bluca** [comment]
As you can see from the linked commit, there is already a build time check for the `sd_pidfd_get_session` function, and a fallback is used if not available, which would be your case. So the issue is somewhere else, and you'll need to debug further and find out what exactly is failing and where, as this is not really actionable as it is.

**Comment by 0strodamus** [comment]
Thanks for the reply.

I'm not a developer, so maybe I'm missing the fallback you're referring to. I see an if/else statement in the linked commit where sd_pid_get_session appears to be the fallback to sd_pidfd_get_session, however they're wrapped in a HAVE_LIBSYSTEMD ifdef/endif so this whole section of code should not be included when I compile because my system does not have systemd or elogind installed. I'm using ConsoleKit 1.26 with OpenRC/sysV init.

I have spent a lot of time looking over the code and I don't know how to debug this any further. I'm content to keep hacking meson.build, although looking at the commits made to polkit-125, I may stay on 123 and backport any future changes that aren't targeting systems with elogind/systemd.

If this is still not actionable, please close my issue report. Thanks again for taking the time to reply and all the best to you and the polkit project.

**Comment by thesamesam** [comment]
We had a report of something similar with elogind in Gentoo (https://bugs.gentoo.org/934314) but I haven't dug into it more (yet).

**Comment by RomainNaour** [comment]
Hello,

We also noticed this regression with [PIDFD]([https://github.com/polkit-org/polkit/commit/7f0d792323cdca70b3d581cc9ed54df3d844a637) using SYSV init.

The issue is reproduced with polkit 124 and 125 (and master)

https://gitlab.com/buildroot.org/buildroot/-/jobs/8199992596

**Comment by vmihalko** [comment]
I was unable to reproduce the issue with elogind-255.5 on the Gentoo LiveGUI OpenRC image. Here’s the process I followed:

```
git clone https://github.com/polkit-org/polkit.git; cd polkit
git checkout 126
git cherry-pick 55ee1b70456eca8281dda9612c485c619122f202 #fix elogind support
meson setup -D session_tracking=elogind build
meson compile -C build
meson install -C build
```

I checked if the ["Sleep," "Power off," and "Restart" buttons were grayed out](https://bugs.gentoo.org/934314) or tested `pkexec gedit`, and everything worked fine.

I also tried various versions of polkit: 12{3,4,5,6}. Could you please provide a reproducer for this issue?

**Comment by 0strodamus** [comment]
Sure. Uninstall elogind and systemd. Install [ConsoleKit 1.2.6](https://github.com/ConsoleKit2/ConsoleKit2/releases) and try again using polkit build option "session_tracking=ConsoleKit".

**Comment by thesamesam** [comment]
That obviously isn't a reproducer for the Gentoo case given we don't package ConsoleKit at this time. You're free to give your own reproduction instructions for your case, but please don't be glib either.

**Comment by 0strodamus** [comment]
My apologies if my reply is perceived as glib. That wasn't my intention.

**Comment by vmihalko** [comment]
More reproducers are always welcome. So, @thesamesam, do you have a reproducer for Gentoo? And @0strodamus, could you share which OS you're using, or suggest one for testing polkit with non-systemd and consolekit?

**Comment by 0strodamus** [comment]
The OS I'm running started out a very long time ago as Arch Linux, but I removed systemd and replaced it with sysvinit/OpenRC/Consolekit. Due to removing systemd I have to compile a ton of stuff locally that I can't pull directly from Arch (including polkit). I would say that [Artix Linux](https://artixlinux.org/) is the closest distro to what I'm doing locally, although they're using elogind instead of Consolekit and OpenRC init instead of SysVinit.

As far as a distro that you could test on, the only one I've found that offers Consolekit as a pre-packaged option to elogind is [Devuan 5 Daedalus](https://files.devuan.org/devuan_daedalus/Release_notes.txt). [AntiX Linux](https://antixlinux.com/) could be a second option, but it appears you would have to compile Consolekit locally.

**Comment by jrybar-rh** [comment]
Currently, I cannot imagine a capacity that could look into finding an environment this specific to run into such a rare  reproducer. As a workaround for the time being, is using elogind a viable option for you, @0strodamus ?

**Comment by jrybar-rh** [comment]
@0strodamus Also (just shooting blindly), I read on [Arch Wiki](https://wiki.archlinux.org/title/ConsoleKit) that running ConsoleKit requires `consolekit-polkit` package. Do you build that too?

**Comment by 0strodamus** [comment]
I agree with you about the difficulty in finding an environment similar to mine to reproduce. I couldn't find any distro that ships with sysvinit and with Consolekit as the defaults. As I wrote earlier, I'm content to keep hacking meson.build as a work-around which is my preference over using elogind.

Yes, [polkit-consolekit](https://aur.archlinux.org/packages/polkit-consolekit) is what I'm building locally.

Hopefully if a reproducer can be more easily found in @thesamesam or @RomainNaour cases, the fix will also solve the issue for mine...

**Comment by thesamesam** [comment]
I've asked the users on the bug on our side to try craft something.
